### PR TITLE
[SPARK-40334][PS] Implement `GroupBy.prod`

### PR DESCRIPTION
**Pull request by ayudovin**
### What changes were proposed in this pull request?
Implement `GroupBy.prod`

### Why are the changes needed?
for API coverage

### Does this PR introduce _any_ user-facing change?
yes, the new API
```
df = ps.DataFrame({'A': [1, 1, 2, 1, 2],
                            'B': [np.nan, 2, 3, 4, 5],
                            'C': [1, 2, 1, 1, 2],
                            'D': [True, False, True, False, True]})

        Groupby one column and return the prod of the remaining columns in
        each group.

        df.groupby('A').prod()
             B  C  D
        A
        1  8.0  2  0
        2  15.0 2  11

        df.groupby('A').prod(min_count=3)
             B  C   D
        A
        1  NaN  2  0
        2  NaN NaN  NaN
```
        

### How was this patch tested?
added UT

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @zhengruifeng @xinrong-meng @itholic FYI

**Comment by zhengruifeng** [comment]
also cc @Yikun

File: python/pyspark/pandas/groupby.py [hunk 1]
```diff
@@ -993,6 +993,98 @@ def nth(self, n: int) -> FrameLike:
...
+                else:
+                    stat_exprs.append(F.product(column).alias(f"{label[0]}"))
+
+                stat_exprs.append(F.count(column).alias(f"{label[0]}_count"))
+
+            sdf = sdf.groupby(*groupkey_names).agg(*stat_exprs)
+
+            if min_count > 0:
+                for label in psdf._internal.column_labels:
+                    sdf = sdf.withColumn(f"{label[0]}", F.when(F.col(f"{label[0]}_count").__ge__(min_count),
```
Comment (Yikun) [inline python/pyspark/pandas/groupby.py:1073]: we usually choose to create a temporary column with a special name and verify (`verify_temp_column_name`), and then drop it.

see below as reference:
https://github.com/apache/spark/blob/master/python/pyspark/pandas/groupby.py#L1024-L1032

File: python/pyspark/pandas/groupby.py [hunk 2]
```diff
@@ -993,6 +993,98 @@ def nth(self, n: int) -> FrameLike:
...
+                psser = psdf._psser_for(label)
+                column = psser._dtype_op.nan_to_null(psser).spark.column
+                data_type = psser.spark.data_type
+
+                if isinstance(data_type, IntegralType):
+                    stat_exprs.append(F.product(column).cast(data_type).alias(f"{label[0]}"))
+                else:
+                    stat_exprs.append(F.product(column).alias(f"{label[0]}"))
+
+                stat_exprs.append(F.count(column).alias(f"{label[0]}_count"))
```
Comment (Yikun) [inline python/pyspark/pandas/groupby.py:1067]: Looks like we also don't need this helper column if min_count==0, right?

File: python/pyspark/pandas/groupby.py [hunk 3]
```diff
@@ -993,6 +993,98 @@ def nth(self, n: int) -> FrameLike:
...
+        ----------
+        numeric_only : bool, default False
+        Include only float, int, boolean columns. If None, will attempt to use
+        everything, then use only numeric data.
+
+        min_count: int, default 0
+        The required number of valid values to perform the operation.
+        If fewer than min_count non-NA values are present the result will be NA.
+
+        .. versionadded:: 3.4.0
```
[the remark this answers is not in the document]
Comment (Yikun) [reply python/pyspark/pandas/groupby.py:1010]: You could also generate and see final doc in local: 
https://spark.apache.org/docs/latest/api/python/development/contributing.html#contributing-documentation-changes

File: python/pyspark/pandas/groupby.py (hunk 2 above)
Comment (ayudovin) [reply python/pyspark/pandas/groupby.py]: yes, you're right

File: python/pyspark/pandas/groupby.py [hunk 4]
```diff
@@ -44,6 +43,7 @@
 )
 import warnings
 
+import numpy as np
```
Comment (zhengruifeng) [inline python/pyspark/pandas/groupby.py:46]: `numpy` in the docstring was imported in `def _test()`, so no need to import it here

File: python/pyspark/pandas/groupby.py [hunk 5]
```diff
@@ -993,6 +994,101 @@ def nth(self, n: int) -> FrameLike:
...
+            Include only float, int, boolean columns. If None, will attempt to use
+            everything, then use only numeric data.
+
+        min_count: int, default 0
+            The required number of valid values to perform the operation.
+            If fewer than min_count non-NA values are present the result will be NA.
+
+        Returns
+        -------
+        pyspark.pandas.Series or pyspark.pandas.DataFrame
```
Comment (zhengruifeng) [inline python/pyspark/pandas/groupby.py:1015]:
```suggestion
        Series or DataFrame
```

**Comment by zhengruifeng** [comment]
```
Oh no! 💥 💔 💥
2 files would be reformatted, 352 files would be left unchanged.
Please run 'dev/reformat-python' script.
1
Error: Process completed with exit code 1.
```

`lint-python` failed, you would install related packages in your env via `pip install -r dev/requirements.txt `, and then reformat the python files via `dev/reformat-python` and check the lint locally via `dev/lint-python`

File: python/pyspark/pandas/groupby.py [hunk 6]
```diff
@@ -993,6 +994,101 @@ def nth(self, n: int) -> FrameLike:
...
+        groupkey_names = [SPARK_INDEX_NAME_FORMAT(i) for i in range(len(self._groupkeys))]
+        internal, agg_columns, sdf = self._prepare_reduce(
+            groupkey_names=groupkey_names,
+            accepted_spark_types=(NumericType, BooleanType),
+            bool_to_numeric=True,
+        )
+
+        psdf: DataFrame = DataFrame(internal)
+        if len(psdf._internal.column_labels) > 0:
+            tmp_count_column = "__tmp_%s_count_col__"
```
Comment (Yikun) [inline python/pyspark/pandas/groupby.py:1056]:
```suggestion
            tmp_count_column = verify_temp_column_name(psdf, "__tmp_%s_count_col__")
```

You might want to verify column to aovid pontential column name conflict.

File: python/pyspark/pandas/groupby.py [hunk 7]
```diff
@@ -18,7 +18,6 @@
 """
 A wrapper for GroupedData to behave similar to pandas GroupBy.
 """
-
```
Comment (Yikun) [inline python/pyspark/pandas/groupby.py:21]: unrelated change

File: python/pyspark/pandas/groupby.py [hunk 8]
```diff
@@ -1052,10 +1052,10 @@ def prod(self, numeric_only: Optional[bool] = True, min_count: int = 0) -> Frame
 
         psdf: DataFrame = DataFrame(internal)
         if len(psdf._internal.column_labels) > 0:
-            tmp_count_column = verify_temp_column_name(psdf, "__tmp_%s_count_col__")
 
             stat_exprs = []
             for label in psdf._internal.column_labels:
+                tmp_count_column = verify_temp_column_name(sdf, "__tmp_%s_count_col__" % label[0])
```
Comment (itholic) [inline python/pyspark/pandas/groupby.py:1058]: nit: use `tmp_count_column_name` instead to make it more explicit ?

File: python/pyspark/pandas/groupby.py [hunk 9]
```diff
@@ -993,6 +993,105 @@ def nth(self, n: int) -> FrameLike:
...
+            Include only float, int, boolean columns. If None, will attempt to use
+            everything, then use only numeric data.
+
+        min_count: int, default 0
+            The required number of valid values to perform the operation.
+            If fewer than min_count non-NA values are present the result will be NA.
+
+        Returns
+        -------
+        Series or DataFrame
```
Comment (itholic) [inline python/pyspark/pandas/groupby.py:1014]: Let's add a sub-description as pandas does.

```suggestion
        Series or DataFrame
            Computed prod of values within each group.
```

File: python/pyspark/pandas/groupby.py [hunk 10]
```diff
@@ -993,6 +993,105 @@ def nth(self, n: int) -> FrameLike:
...
+        --------
+        pyspark.pandas.Series.groupby
+        pyspark.pandas.DataFrame.groupby
+
+        Examples
+        --------
+        >>> df = ps.DataFrame({'A': [1, 1, 2, 1, 2],
+        ...                    'B': [np.nan, 2, 3, 4, 5],
+        ...                    'C': [1, 2, 1, 1, 2],
+        ...                    'D': [True, False, True, False, True]})
```
Comment (itholic) [inline python/pyspark/pandas/groupby.py:1026]: nit: formatting


```suggestion
        >>> df = ps.DataFrame(
        ...     {
        ...         "A": [1, 1, 2, 1, 2],
        ...         "B": [np.nan, 2, 3, 4, 5],
        ...         "C": [1, 2, 1, 1, 2],
        ...         "D": [True, False, True, False, True],
        ...     }
        ... )
```

File: python/pyspark/pandas/groupby.py [hunk 11]
```diff
@@ -993,6 +993,105 @@ def nth(self, n: int) -> FrameLike:
...
+            for label in psdf._internal.column_labels:
+                tmp_count_column = verify_temp_column_name(sdf, "__tmp_%s_count_col__" % label[0])
+                psser = psdf._psser_for(label)
+                column = psser._dtype_op.nan_to_null(psser).spark.column
+                data_type = psser.spark.data_type
+
+                if isinstance(data_type, IntegralType):
+                    stat_exprs.append(F.product(column).cast(data_type).alias(label[0]))
+                else:
+                    stat_exprs.append(F.product(column).alias(label[0]))
```
Comment (itholic) [inline python/pyspark/pandas/groupby.py:1066]: What about defining a `label[0]` as a variable since it's used in multiple places ?

File: python/pyspark/pandas/groupby.py [hunk 12]
```diff
@@ -993,6 +993,105 @@ def nth(self, n: int) -> FrameLike:
...
+
+            sdf = sdf.groupby(*groupkey_names).agg(*stat_exprs)
+
+            if min_count > 0:
+                for label in psdf._internal.column_labels:
+                    tmp_count_column = "__tmp_%s_count_col__"
+                    sdf = sdf.withColumn(
+                        label[0],
+                        F.when(
+                            F.col(tmp_count_column % label[0]).__ge__(min_count), F.col(label[0])
```
Comment (itholic) [inline python/pyspark/pandas/groupby.py:1079]: ditto ?

**Review by itholic — COMMENTED** [review]
Looks good except several minor comments.

File: python/pyspark/pandas/groupby.py (hunk 11 above)
Comment (ayudovin) [reply python/pyspark/pandas/groupby.py]: It makes sense

File: python/pyspark/pandas/groupby.py [hunk 13]
```diff
@@ -993,6 +993,115 @@ def nth(self, n: int) -> FrameLike:
...
+
+                if isinstance(data_type, IntegralType):
+                    stat_exprs.append(F.product(column).cast(data_type).alias(label_name))
+                else:
+                    stat_exprs.append(F.product(column).alias(label_name))
+
+                if min_count > 0:
+                    stat_exprs.append(F.count(column).alias(tmp_count_column_name))
+
+            sdf = sdf.groupby(*groupkey_names).agg(*stat_exprs)
```
Comment (zhengruifeng) [inline python/pyspark/pandas/groupby.py:1079]: Can we simplify the logic like this?

```suggestion
            for label in psdf._internal.column_labels:
                psser = psdf._psser_for(label)
                scol = psser._dtype_op.nan_to_null(psser).spark.column
                prod_scol = F.product(scol)
                if min_count > 0:
                    prod_scol = F.when(F.count(scol) < min_count, F.lit(None)).otherwise(F.product(scol))
                else:
                    prod_scol = F.product(scol)
                
                stat_exprs.append(
                    prod_scol.alias(psser._internal.data_spark_column_names[0])
                )
            sdf = sdf.groupby(*groupkey_names).agg(*stat_exprs)
```

Comment (zhengruifeng) [inline python/pyspark/pandas/groupby.py:1079]: Can we simplify the logic like this?

```suggestion
            for label in psdf._internal.column_labels:
                psser = psdf._psser_for(label)
                scol = psser._dtype_op.nan_to_null(psser).spark.column
                prod_scol = F.product(scol)
                if min_count > 0:
                    prod_scol = F.when(F.count(scol) < min_count, F.lit(None)).otherwise(F.product(scol))
                else:
                    prod_scol = F.product(scol)
                
                stat_exprs.append(
                    prod_scol.alias(psser._internal.data_spark_column_names[0])
                )
            sdf = sdf.groupby(*groupkey_names).agg(*stat_exprs)
```

File: python/pyspark/pandas/groupby.py [hunk 14]
```diff
@@ -993,6 +993,115 @@ def nth(self, n: int) -> FrameLike:
...
+            for label in psdf._internal.column_labels:
+                label_name = label[0]
+                tmp_count_column_name = verify_temp_column_name(
+                    sdf, "__tmp_%s_count_col__" % label_name
+                )
+                psser = psdf._psser_for(label)
+                column = psser._dtype_op.nan_to_null(psser).spark.column
+                data_type = psser.spark.data_type
+
+                if isinstance(data_type, IntegralType):
```
Comment (zhengruifeng) [inline python/pyspark/pandas/groupby.py:1071]: for `IntegralType `, maybe always returns `LongType`?

File: python/pyspark/pandas/groupby.py [hunk 15]
```diff
@@ -993,6 +993,115 @@ def nth(self, n: int) -> FrameLike:
...
+            Computed prod of values within each group.
+
+        See Also
+        --------
+        pyspark.pandas.Series.groupby
+        pyspark.pandas.DataFrame.groupby
+
+        Examples
+        --------
+        >>> df = ps.DataFrame(
```
Comment (Yikun) [inline python/pyspark/pandas/groupby.py:1024]:
```suggestion
        >>> import numpy as np
        >>> df = ps.DataFrame(
```

Let's make it as self-contained examples, it means users can copy/paste to python shell directly, see also: SPARK-40005

File: python/pyspark/pandas/groupby.py (hunk 7 above)
Comment (Yikun) [reply python/pyspark/pandas/groupby.py:21]: Friendly note, looks like it is not recovered yet.

File: python/pyspark/pandas/groupby.py (hunk 13 above)
Comment (ayudovin) [reply python/pyspark/pandas/groupby.py]: yes, we can, I applied changes

File: python/pyspark/pandas/groupby.py (hunk 14 above)
Comment (ayudovin) [reply python/pyspark/pandas/groupby.py]: What do you mean by `always returns LongType`? 
Always cast the result of F.product() to LongType?

Comment (ayudovin) [reply python/pyspark/pandas/groupby.py]: Do you mean we won't have a case when data_type is not the LongType?

File: python/pyspark/pandas/groupby.py [hunk 16]
```diff
@@ -993,6 +993,106 @@ def nth(self, n: int) -> FrameLike:
...
+                column = psser._dtype_op.nan_to_null(psser).spark.column
+                data_type = psser.spark.data_type
+                aggregating = (
+                    F.product(column).cast(data_type)
+                    if isinstance(data_type, IntegralType)
+                    else F.product(column)
+                )
+
+                if min_count > 0:
+                    prod_scol = F.when(F.count(column) < min_count, F.lit(None)).otherwise(
```
Comment (zhengruifeng) [inline python/pyspark/pandas/groupby.py:1074]:
```suggestion
                    prod_scol = F.when(F.count(F.when(~F.isnull(column), F.lit(0))) < min_count, F.lit(None)).otherwise(
```

sorry, I forgot to filter out invalid values in previous suggestion

**Review by zhengruifeng — APPROVED** [review]
LGTM otherwise

File: python/pyspark/pandas/groupby.py [hunk 17]
```diff
@@ -993,6 +993,106 @@ def nth(self, n: int) -> FrameLike:
...
+        psdf: DataFrame = DataFrame(internal)
+        if len(psdf._internal.column_labels) > 0:
+
+            stat_exprs = []
+            for label in psdf._internal.column_labels:
+                psser = psdf._psser_for(label)
+                column = psser._dtype_op.nan_to_null(psser).spark.column
+                data_type = psser.spark.data_type
+                aggregating = (
+                    F.product(column).cast(data_type)
```
[the remark this answers is not in the document]
Comment (ayudovin) [reply python/pyspark/pandas/groupby.py]: Yep, that makes sense, I added it

**Comment by zhengruifeng** [comment]
Merged into master, thank you @ayudovin for working on it!
